src/cpp/runner/main.cpp

```diff
@@ -9,26 +10,144 @@
 
 using namespace std;
 
-/*GMMOutput readInputData() {
+GMMInput read_input_data(const string& input_file, const bool replicate_point)
+{
+    GMMInput input;
+
+    // Read instance
+    read_gmm_instance(input_file, &input.d, &input.k, &input.n,
+        input.alphas, input.means, input.icf, input.x, input.wishart, replicate_point);
+
+    return input;
+}
+
+typedef  void (ITest<GMMInput, GMMOutput>::* test_member_function) (int);
+
+inline void call_member_function(unique_ptr<ITest<GMMInput, GMMOutput>>& ptr_to_object, const test_member_function ptr_to_member, const int arg1) {
+    (*(ptr_to_object).*(ptr_to_member))(arg1);
+}
+
+double measure_shortest_time(const double minimum_measurable_time, const int nruns, const double time_limit, unique_ptr<ITest<GMMInput, GMMOutput>>& test, const test_member_function func)
+{
+    vector<double> samples;
+    double total_time = 0;
+    auto repeats = 1;
+    for (;; repeats *= 2)
+    {
+        auto t1 = high_resolution_clock::now();
+        call_member_function(test, func, repeats);
+        auto t2 = high_resolution_clock::now();
+        const auto current_run_time = duration_cast<duration<double>>(t2 - t1).count();
+        if (current_run_time > minimum_measurable_time)
+        {
+            samples.push_back(current_run_time / repeats);
+            total_time += current_run_time;
+            break;
+        }
+    }
+
+    for (auto run = 1; (run < nruns) && (total_time < time_limit); run++)
+    {
+        auto t1 = high_resolution_clock::now();
+        call_member_function(test, func, repeats);
+        auto t2 = high_resolution_clock::now();
+        const auto current_run_time = duration_cast<duration<double>>(t2 - t1).count();
+        samples.push_back(current_run_time / repeats);
+        total_time += current_run_time;
+    }
+
+    const auto min_sample = *std::min_element(std::begin(samples), std::end(samples));
+
+    return min_sample;
+}
+
+std::string filepath_to_basename(const std::string& filepath)
+{
+    const auto filename = filepath.substr(filepath.find_last_of("/\\") + 1);
+    const auto dot = filename.find_last_of('.');
+    auto basename = filename.substr(0, dot);
+
+    return basename;
+}
 
-}*/
+void save_time_to_file(const string& filepath, const double objective_time, const double derivative_time)
+{
+    std::ofstream out(filepath);
+    out << std::scientific << objective_time << "\t" << derivative_time;
+    out.close();
```

Comment:
Probably worth adding a newline, as it is human readable anyway, so it's reasonable to make it human-pleasant.  E.g. I sometimes ``cat gmm_times.txt`` on the commandline.